### PR TITLE
MCR-2118 add support external IIIF manifests to viewer

### PR DESCRIPTION
**Pull request by eckstueck**
added support for external IIIF manifests to viewer

[Link to jira](https://mycore.atlassian.net/browse/MCR-).

### CODE REVIEW

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 9868494c392f4d18f2363903f99b69ec59543b07! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5051-pull_request-878

**Comment by kkrebs** [comment]
in MCR-2118 LTS2019.06 is set as fix version, please use this branch as base

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 9868494c392f4d18f2363903f99b69ec59543b07! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5051-pull_request-878

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/mycore/mycore/pullRequest?prid=4436103&bid=15008166)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts [hunk 1]
```diff
@@ -78,28 +80,45 @@ namespace mycore.viewer.widgets.iiif {
...
+                this.vChapterIdMap.set(chapterObject.id, chapterObject);
+                cans.forEach((can: ICanvas) => {
+                    const childChap = new model.StructureChapter(chapterObject,
+                        chapterObject, this.getIDFromURL(can.id), can.getDefaultLabel());
+                    chapterObject.chapter.push(childChap);
+                    this.vChapterIdMap.set(childChap.id, childChap);
+                });
+            }
+            else {
+                chapterObject = new model.StructureChapter(parent,
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts [hunk 2]
```diff
@@ -78,28 +80,45 @@ namespace mycore.viewer.widgets.iiif {
...
             //     return;
             // }
             //TODO Chaptertype currently not in Manifest
-            const chapterObject = new model.StructureChapter(parent,
-                '', this.getIDFromURL(chapter.id), chapter.getDefaultLabel());
-            // let chapterChildren = chapter.getRanges();
-
-            this.vChapterIdMap.set(chapterObject.id, chapterObject);
+            let chapterObject;
+            if (chapter === undefined || (chapter.getCanvasIds().length === 0 && chapter.getRanges().length === 0)) {
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: [Unexpected use of undefined. (no-undefined)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts [hunk 3]
```diff
@@ -78,28 +80,45 @@ namespace mycore.viewer.widgets.iiif {
...
-                chapterObject.chapter.push(this.processChapter(chapterObject, childChap));
-            });
             return chapterObject;
         }
 
         private getIdFileMap(): MyCoReMap<string, IAnnotation> {
             const map = new MyCoReMap<string, IAnnotation>();
             this.manifestDocument.getSequences()[0].getCanvases().forEach((canvas: ICanvas) => {
                 canvas.getImages().forEach((image: IAnnotation) => {
+                    if (image.id == 'undefined') {
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: [Expected '===' and instead saw '=='. (eqeqeq)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts [hunk 4]
```diff
@@ -37,7 +38,7 @@ namespace mycore.viewer.widgets.iiif {
                 url: manifestDocumentLocation,
                 success: (response: any) => {
                     const manifest = <Manifest> manifesto.create(response);
-                    const builder = new IIIFStructureBuilder(<Manifest> manifest, tilePathBuilder);
+                    const builder = new IIIFStructureBuilder(<Manifest> manifest, tilePathBuilder, imageAPIURL);
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts]: Codacy Issue found: ['IIIFStructureBuilder' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts [hunk 5]
```diff
@@ -48,7 +48,12 @@ namespace mycore.viewer {
 
             settings.filePath = encodeURI(settings.filePath);
 
-            if(settings.webApplicationBaseURL.lastIndexOf("/")==settings.webApplicationBaseURL.length-1){
+            if (settings.filePath.indexOf('_derivate_') > 0) {
+                settings.filePath = settings.filePath.replace('/', '%2F');
+            }
+
+            if(settings.webApplicationBaseURL != "undefined"
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts]: Codacy Issue found: [Expected '!==' and instead saw '!='. (eqeqeq)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts [hunk 6]
```diff
@@ -78,28 +80,45 @@ namespace mycore.viewer.widgets.iiif {
...
-            // let chapterChildren = chapter.getRanges();
-
-            this.vChapterIdMap.set(chapterObject.id, chapterObject);
+            let chapterObject;
+            if (chapter === undefined || (chapter.getCanvasIds().length === 0 && chapter.getRanges().length === 0)) {
+                chapterObject = new model.StructureChapter(parent,
+                    '', 'LOG_0', this.manifestDocument.getDefaultLabel());
+                this.vChapterIdMap.set(chapterObject.id, chapterObject);
+                cans.forEach((can: ICanvas) => {
+                    const childChap = new model.StructureChapter(chapterObject,
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/mycore/mycore/pullRequest?prid=4436103&bid=15008166)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts (hunk 4 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts]: Codacy Issue found: ['IIIFStructureBuilder' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 1 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 2 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: [Unexpected use of undefined. (no-undefined)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 6 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/mycore/mycore/pullRequest?prid=4436103&bid=15008166)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts (hunk 4 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IviewIIIFProvider.ts:41]: Codacy Issue found: ['IIIFStructureBuilder' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 6 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts:94]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 1 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts:101]: Codacy Issue found: ['model' is not defined. (no-undef)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts [hunk 7]
```diff
@@ -48,7 +48,12 @@ namespace mycore.viewer {
 
             settings.filePath = encodeURI(settings.filePath);
 
-            if(settings.webApplicationBaseURL.lastIndexOf("/")==settings.webApplicationBaseURL.length-1){
+            if (settings.filePath.indexOf('_derivate_') > 0) {
+                settings.filePath = settings.filePath.replace('/', '%2F');
+            }
+
+            if(settings.webApplicationBaseURL !== "undefined"
+                && settings.webApplicationBaseURL.lastIndexOf("/") === settings.webApplicationBaseURL.length - 1){
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts:56]: Codacy Issue found: [" should be '](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts [hunk 8]
```diff
@@ -48,7 +48,12 @@ namespace mycore.viewer {
 
             settings.filePath = encodeURI(settings.filePath);
 
-            if(settings.webApplicationBaseURL.lastIndexOf("/")==settings.webApplicationBaseURL.length-1){
+            if (settings.filePath.indexOf('_derivate_') > 0) {
+                settings.filePath = settings.filePath.replace('/', '%2F');
+            }
+
+            if(settings.webApplicationBaseURL !== "undefined"
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts:55]: Codacy Issue found: [missing whitespace](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/widgets/chapter/MobileChapterTreeView.ts [hunk 9]
```diff
@@ -38,6 +38,9 @@ namespace mycore.viewer.widgets.chaptertree {
         private static LEVEL_MARGIN = 15;
 
         addNode(parentId:string, id:string, label:string, childLabel: string, expandable:boolean) {
+            if(label === childLabel) {
```
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/widgets/chapter/MobileChapterTreeView.ts:41]: Codacy Issue found: [missing whitespace](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts (hunk 7 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts:56]: Codacy Issue found: [missing whitespace](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts (hunk 2 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/iiif/widgets/IIIFStructureBuilder.ts:89]: Codacy Issue found: [Unexpected use of undefined. (no-undefined)](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

File: mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts (hunk 8 above)
Comment (adlerre) [inline mycore-viewer/src/main/typescript/modules/base/MyCoReViewerSettings.ts:55]: Codacy Issue found: [" should be '](https://app.codacy.com/app/MyCoRe/mycore/pullRequest?prid=4436103)

**Review by kkrebs — APPROVED** [review]

**Pull request merged**